# Index/Query scripts

This is Python source folder of the hspsdb project, for indexing and querying
sequence similarity search results.

## Search results in tabular format

* [indexHSPs.py](indexHSPs.py): Index similarity search results
 in tabular format, e.g. generated by [LAMBDA](https://github.com/seqan/lambda),
 [BLASTX](http://blast.ncbi.nlm.nih.gov),
 or [DIAMOND](https://github.com/bbuchfink/diamond)

 ```bash
 ./hspsdb/indexHSPs.py index ~/study1/data/cutadapt-blastx/sample1.R1.m8\
  sample1 study1-blastx --pair=1
 ```

* [queryHSPs.py](queryHSPs.py): Query indexed similarity search results,
  find most represented genes and present them using PivotTable.js
  
 ```bash
./hspsdb/queryHSPs.py topgenes  study1-diamond  study1-topgenes\
 --bitscore 120
chrome study1-diamond-topgenes.html
# If output file name not specified it is set to the name of the study    
./hspsdb/queryHSPs.py topgenes studyk --bitscore 60
chrome studyk.html
 ```

## Search results in SAM/BAM format

* [indexbam.py](indexbam.py): Index SAM BAM files with MongoDB,
  _at early stages of development_

  BAM files reading was initially based on
  [pybam](https://github.com/JohnLonginotto/pybam) library,
  latest version reads both BAM and SAM files by calling
  [pysam](https://github.com/pysam-developers/pysam) library functions
