HSPs-db project is an effort to develop index and query tools for
sequence similarity search results.

Although project supports NCBI-BLAST xml/json formats and SAM/BAM
formats, its current focus is on BLAST tabular results format.

Initally only Elasticsearch indexes were supported, later
MongoDB support added. Current focus on tabular search results
is based on MongoDB indexes only.

## History and project structure

Initial effort of the project was for BLAST xml/json files.
Later we wanted to support for SAM Sequence Alignment/Map format,
using Java [HTSJDK](https://github.com/samtools/htsjdk) library.
Later we have found pybam and pysam libraries as alternatives for reading SAM/BAM
files and have started adding Python code to the project.
For this reason project is not a pure Java, Python or JavaScript project,
and have project files for all these 3 languages; _pom.xml_, _setup.py_,
_package.json_.


## Installation

Download hspsdb project source code and install required libraries:

```bash
git clone https://bitbucket.org/hspsdb/nosql-biosets.git ./nosqlbiosets
cd nosqlbiosets
pip install -r requirements.txt --user
cd ..
git clone https://bitbucket.org/hspsdb/hspsdb-indexer.git ./hspsdb
cd hspsdb
pip install -r requirements.txt --user
```

## Indexing and querying tabular results

* Install [MongoDB](https://www.mongodb.com)

* [indexHSPs.py](./hspsdb/indexHSPs.py): Index similarity search results saved
 in tabular format, e.g. generated by [LAMBDA](https://github.com/seqan/lambda),
 [BLASTX](http://blast.ncbi.nlm.nih.gov),
 or [DIAMOND](https://github.com/bbuchfink/diamond)

Example command lines to index LAMBDA search results of sequencing reads from
two samples

 ```bash 
./hspsdb/indexHSPs.py index ~/studyk/cutadapt-lambda2/sample1.R1.m8 sample1 studyk --pair=1
./hspsdb/indexHSPs.py index ~/studyk/cutadapt-lambda2/sample1.R2.m8 sample1 studyk --pair=2
./hspsdb/indexHSPs.py index ~/studyk/cutadapt-lambda2/sample2.R1.m8 sample2 studyk --pair=1
./hspsdb/indexHSPs.py index ~/studyk/cutadapt-lambda2/sample2.R2.m8 sample2 studyk --pair=2  
 ```

* [queryHSPs.py](./hspsdb/queryHSPs.py): Query indexed similarity search results,
  find most represented genes and present them using PivotTable.js
  
  
 ```bash
./hspsdb/queryHSPs.py topgenes  studyk  studyk-topgenes\
 --bitscore 120
chrome study1-diamond-topgenes.html
# If output file name not specified it is set to the name of the study    

./hspsdb/queryHSPs.py topgenes studyk --bitscore 60

chrome studyk.html

 ```

## Indexing results in BLAST xml, json formats

Following steps describe how to index BLAST result files in xml/json formats. 
We will add new sections when the index scripts for SAM/BAM files and tabular
result files reach to a level of maturity.

* Install [Elasticsearch](https://www.elastic.co)

  If you are new to Elasticsearch
  the easiest way is to [download Elasticsearch](
  https://www.elastic.co/downloads/elasticsearch) with the TAR option (~30M).
  After extracting the tar file just `cd` to your Elasticsearch folder
  and run `./bin/elasticsearch` command.

* Choose an Elasticsearch index name for your BLAST results
  and use [init-index.sh](scripts/init-index.sh)
  script to initialize your index

* If you already have your BLAST outputs in json format use
  [index.sh](scripts/index.sh)
  or [index-folder.sh](scripts/files2index.sh) scripts
  to index your output files.
  If you do not have any BLAST results in json format and if you first want to
  see how HSPs-db is working,
  then you can use the sample results we have in the testdb folder.
  Call [files2index.sh](scripts/files2index.sh) script with argument `./testdb`
  followed by the index name you choose earlier, and the URL of your Elasticsearch
  server.
  If you already have outputs in BLAST archive format you can use
  `blast_formatter` command to regenerate your outputs in json format.
  We want to implement scripts that maps existing XML outputs to json format
  but this has not been done yet

* After you indexed your BLAST results you can install
 [HSPs-db web interface](https://github.com/uludag/hspsdb-webcode)
  for querying and visualizing indexed BLAST results

* For new BLAST searches it is best to produce outputs first in BLAST archive format
  then use `blast_formatter` to generate the outputs you normally read
  and the json output for indexing

* This repository also includes Node.js client scripts for EMBL-EBI and NCBI,
  BLAST services; [ebi-sss-client.js](scripts/ebi-sss-client.js),
  [ncbi-sss-client.js](scripts/ncbi-sss-client.js).
  Modules required by the scripts can be installed
  by running `npm install` from project root folder

## Similar work

* https://github.com/daler/gffutils:
  "GFF and GTF files are loaded into SQLite3 databases,
  allowing much more complex manipulation of hierarchical features
  (e.g., genes, transcripts, and exons) than is possible with plain-text methods
  alone"
  
  _nosql-biosets project doesn't yet have a level of maturity comparable
   to the gffutils library_.

* We can say we want to achieve what the [MultiQC](http://multiqc.info) project
  has already achieved; "A modular tool to aggregate results from bioinformatics
  analyses across many samples into a single report", _with more dynamic reports_.

* We can also say HSPs-db project has some similarity to the
 [SeQC](https://github.com/JohnLonginotto/SeQC) project whic is maintained
 by the developer of the 'pybam' library which we use for indexing BAM files
 
 * https://www.monetdb.org/Documentation/Extensions/LifeScience/load 

## Notes

In a separate repository we develop web interfaces for the indexed results,
[hspsdb-webcode](https://github.com/uludag/hspsdb-webcode).
Latest version of the web interface for BLAST results in xml/json outputs
can be seen on a [test server](http://hspsdb-test.herokuapp.com)
which is connected to an Elasticsearch server with a small set of sample
BLAST results.

## Copyright

HSPs-db project has been developed
at King Abdullah University of Science and Technology,
[http://www.kaust.edu.sa](http://www.kaust.edu.sa)

HSPs-db project is licensed with MIT license.
If you would like to support the project
with selecting a different license please let us know by creating an issue
on github project page.
We will help you with contacting the relavant bodies of KAUST.
