# HSPs-db; index and query tools for sequence similarity search results

HSPs-db project is an effort to develop index and query tools for
sequence similarity search results.

Although project supports BLAST xml/json formats and SAM/BAM
formats, its current focus is on tabular results

## Indexing and querying tabular results

* [indexHSPs.py](../hspsdb/indexHSPs.py): Index similarity search results saved
 in tabular format, e.g. generated by [LAMBDA](https://github.com/seqan/lambda),
 [BLASTX](http://blast.ncbi.nlm.nih.gov),
 or [DIAMOND](https://github.com/bbuchfink/diamond)

Example command lines to index LAMBDA search results of sequencing reads from
two samples

 ```bash 
./hspsdb/indexHSPs.py index ~/studyk/cutadapt-lambda2/sample1.R1.m8 sample1 studyk --pair=1
./hspsdb/indexHSPs.py index ~/studyk/cutadapt-lambda2/sample1.R2.m8 sample1 studyk --pair=2
./hspsdb/indexHSPs.py index ~/studyk/cutadapt-lambda2/sample2.R1.m8 sample2 studyk --pair=1
./hspsdb/indexHSPs.py index ~/studyk/cutadapt-lambda2/sample2.R2.m8 sample2 studyk --pair=2  
 ```

* [queryHSPs.py](../hspsdb/queryHSPs.py): Query indexed similarity search results,
  find most represented genes and present them using PivotTable.js
  
  
 ```bash
./hspsdb/queryHSPs.py topgenes  studyk  studyk-topgenes\
 --bitscore 120
chrome study1-diamond-topgenes.html
# If output file name not specified it is set to the name of the study    

./hspsdb/queryHSPs.py topgenes studyk --bitscore 60

chrome studyk.html

 ```
